File: app/src/main/java/com/mathgate/MathAcademyGateApi.kt
package com.mathgate

import java.io.IOException
import java.net.SocketTimeoutException
import java.net.UnknownHostException

/** Adapts MathAcademyClient to the GateApi contract, mapping every failure to a blocking status. */
class MathAcademyGateApi(
    private val client: MathAcademyClient,
    private val credentials: () -> Credentials?,
    private val clock: () -> Long = System::currentTimeMillis,
) : GateApi {

    override fun fetch(request: GateRequest): GateStatus {
        val creds = credentials() ?: return GateStatus.AuthFailed("no_credentials (sign in from the MathGate app)")
        return try {
            val tasks = client.fetchRecentTasks(creds, request.zone)
            val summary = TaskSummariser.summarise(tasks, request.periodStartMillis)
            MgLog.d("summary: ${summary.tasksCompleted} tasks, ${summary.xp} XP since period start (of ${tasks.length()} recent)")
            if (summary.isDone) {
                GateStatus.Done(request.nextResetMillis, summary.tasksCompleted, summary.xp)
            } else {
                GateStatus.NotDone(request.periodStartMillis, request.nextResetMillis, clock())
            }
        } catch (e: MathAcademyClient.AuthFailedException) {
            MgLog.w("auth failed: ${e.reason}")
            GateStatus.AuthFailed(e.reason)
        } catch (e: MathAcademyClient.UnexpectedResponseException) {
            MgLog.w("unexpected response: ${e.detail}")
            GateStatus.UnexpectedResponse(e.detail)
        } catch (e: IOException) {
            MgLog.w("offline: ${e.javaClass.simpleName}: ${e.message}")
            GateStatus.Offline(describe(e))
        }
    }

    private fun describe(e: IOException): String = when (e) {
        is UnknownHostException -> "no internet connection (DNS lookup failed)"
        is SocketTimeoutException -> "Math Academy timed out"
        else -> e.message ?: e.javaClass.simpleName
    }
}

File: ios/MathGateKit/Sources/MathGateKit/MathAcademyGateAPI.swift
import Foundation

/// Turns a Math Academy fetch into a `GateStatus`. Every failure maps to a non-`.done` case,
/// which is what makes the gate fail closed.
public struct MathAcademyGateAPI: GateAPI {
    private let client: MathAcademyClient
    private let secrets: SecretStore
    private let usernameProvider: @Sendable () -> String?
    private let calendar: Calendar

    public init(
        client: MathAcademyClient,
        secrets: SecretStore,
        calendar: Calendar = .autoupdatingCurrent,
        usernameProvider: @escaping @Sendable () -> String?
    ) {
        self.client = client
        self.secrets = secrets
        self.calendar = calendar
        self.usernameProvider = usernameProvider
    }

    public func fetch(_ request: GateRequest) async -> GateStatus {
        guard let username = usernameProvider(), !username.isEmpty else {
            return .authFailed("no_credentials (add your Math Academy account in MathGate)")
        }
        guard let password = secrets.loadPassword(), !password.isEmpty else {
            return .authFailed("no_password (sign in again in MathGate)")
        }
        let creds = Credentials(username: username, password: password)

        do {
            let tasks = try await client.fetchRecentTasks(creds, calendar: calendar)
            let summary = TaskSummariser.summarise(tasks, periodStart: request.periodStart)
            if summary.isDone {
                return .done(
                    doneUntil: request.nextReset,
                    tasksCompleted: summary.tasksCompleted,
                    xp: summary.xp
                )
            }
            return .notDone(
                periodStart: request.periodStart,
                nextReset: request.nextReset,
                checkedAt: request.now
            )
        } catch let error as MathAcademyClient.ClientError {
            switch error {
            case .authFailed(let reason): return .authFailed(reason)
            case .offline(let message): return .offline(message)
            case .unexpectedResponse(let detail): return .unexpectedResponse(detail)
            }
        } catch {
            return .offline("\(error)")
        }
    }
}
